File: Problem_1.py
"""
Problem : 1

Time Complexity : O(n)
Space Complexity : O(1)

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

Finding the middle of the linked list using slow and fast pointer, then seperating the first two halves
and reversing the second part of linked list. Iterating both the parts from their head
checking if the values match, if yes, then its a palindrome linked list, if they don't, then its not a palindrome linked list

"""

# Palindrome Linked List

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """

        if head==None or head.next==None:
            return head
        # finding the middle point of linked list
        slow=head
        fast=head
        while fast.next!=None and fast.next.next!=None:
            slow=slow.next
            fast=fast.next.next
        
        
        head_l2=slow.next
        slow.next=None

        # reversing the other half of linked list

        prev=None
        curr=head_l2

        while curr!=None:
            temp=curr.next
            curr.next=prev
            prev=curr
            curr=temp
        # print(prev.val,head.val)
        # head at the prev
        if head.val!=prev.val:
            return False

        # iterating from both the ends

        while prev!=None:
            if head.val!=prev.val:
                return False
        
            head=head.next
            prev=prev.next
        return True

File: problem1.cpp
/*

// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
Odd and even length of the list compare values not nodes.


// Your code here along with comments explaining your approach
Find the midpoint using slow and fast pointer
Reverse from slow if even else reverse from slow->next if odd
compare their values until reverse reaches null

*/



#include<iostream>

using namespace std;


//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reversed(ListNode* node){
        ListNode* prev {NULL};
        ListNode* slow{node};
        ListNode* fast{node->next};
        while(fast!= NULL){
            slow->next = prev;
            prev = slow;
            slow = fast;
            fast = fast->next;
        }
        slow->next = prev;
        return slow;
    }
    bool isPalindrome(ListNode* head) {
        ListNode* slow {head};
        ListNode* fast {head};
        if(head == NULL) return false;
        if(head->next == NULL) return true;
        while(fast!= NULL && fast->next!=NULL){
            slow = slow->next;
            fast = fast->next->next;
        }

        if(fast == NULL){
            fast = reversed(slow);
        }
        else if(fast->next == NULL){
            fast = reversed(slow->next);
        }
        slow->next = NULL;
        while(fast!=NULL){
            if(fast->val!=head->val) return false;
            fast = fast->next;
            head = head->next;
        }
        return true;
    }
};